### PR TITLE
*: fix a timezone data race which may cause wrong row data

### LINKED ISSUE #40710: data race in TimeZone causes data-index inconsistency
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Since data race is very hard to reproduce, I will show how it happens in theory.

```
update:
    |
    |----  update one row (use sc.timezone to encode row)
    |
ChildExec (in SetPBColumnsDefaultValue(), set sc.timezone = UTC)
```

 Here is the test case:
```
func TestTimezoneDataRace(t *testing.T) {
	store := testkit.CreateMockStore(t)

	tk := testkit.NewTestKit(t, store)
	tk.MustExec("use test")
	tk.MustExec("CREATE TABLE `c1de839f-4305-4edc-b6bf-463b5890430c` (\n  `1340b953-044b-4cc2-81b7-ed153ade630e` float DEFAULT '1.0469099',\n  `de3381af-3319-4035-985e-a83296fbe929` enum('s059b','f','nt','37n5','9w9pd','omp3m','b','8uwfd','1sv7l','x7j9n','ky8') DEFAULT '1sv7l',\n  `a9b6435c-43a2-4370-9256-eb8b64a590eb` smallint(6) DEFAULT '-10093',\n  `ffbf71c5-8559-40f1-a1f8-5e5e03e69abd` time DEFAULT '08:08:53',\n  `31a80f0f-4b8c-4171-9f06-8d191842541f` decimal(54,22) DEFAULT '84447970762009522232772614140425.0705348716771117376820',\n  `54bac6a1-ada5-436a-8c13-bc6ab738db46` bigint(20) DEFAULT '6183436475535305164',\n  `c37f63a0-7802-48bb-ba0b-7b4c5e2b8097` char(255) DEFAULT 'uv9qaoowi3d2w5po2d5224fm4rmp5o5turz1db1iimcyv8fu8b5kyturpwvvat73ty8oi9qnp76pres1drv9',\n  KEY `89566a55-2db4-4cc0-a094-15d1b09cfa45` (`a9b6435c-43a2-4370-9256-eb8b64a590eb`),\n  KEY `981dd6c3-b1ea-415a-adcf-c84b5432e2ab` (`1340b953-044b-4cc2-81b7-ed153ade630e`),\n  KEY `0d7536fc-4f27-4285-919f-b069a1546ca8` (`a9b6435c-43a2-4370-9256-eb8b64a590eb`,`1340b953-044b-4cc2-81b7-ed153ade630e`,`54bac6a1-ada5-436a-8c13-bc6ab738db46`),\n  KEY `4669c9be-7118-452d-84a2-0213edd38536` (`54bac6a1-ada5-436a-8c13-bc6ab738db46`),\n  KEY `2514744c-9be8-4a49-8b8e-0bf13237665f` (`ffbf71c5-8559-40f1-a1f8-5e5e03e69abd`,`c37f63a0-7802-48bb-ba0b-7b4c5e2b8097`),\n  KEY `4a293299-510c-445e-a857-133497e9b6e3` (`ffbf71c5-8559-40f1-a1f8-5e5e03e69abd`,`54bac6a1-ada5-436a-8c13-bc6ab738db46`),\n  KEY `787fbcf9-da78-4bb2-87f0-056e6d163e99` (`31a80f0f-4b8c-4171-9f06-8d191842541f`),\n  KEY `16e1405b-792b-4109-a5e6-90bb469b1141` (`c37f63a0-7802-48bb-ba0b-7b4c5e2b8097`),\n  KEY `075d34b6-cf58-4f0d-93bc-f98cd821fd70` (`de3381af-3319-4035-985e-a83296fbe929`,`a9b6435c-43a2-4370-9256-eb8b64a590eb`,`1340b953-044b-4cc2-81b7-ed153ade630e`,`ffbf71c5-8559-40f1-a1f8-5e5e03e69abd`,`c37f63a0-7802-48bb-ba0b-7b4c5e2b8097`,`54bac6a1-ada5-436a-8c13-bc6ab738db46`)\n) ENGINE=InnoDB DEFAULT CHARSET=gbk COLLATE=gbk_chinese_ci /*T! SHARD_ROW_ID_BITS=6 */ COMMENT='bdbc9632-58d0-4094-9837-28101a67d77e'\nPARTITION BY HASH (`a9b6435c-43a2-4370-9256-eb8b64a590eb`) PARTITIONS 7;")
	tk.MustExec("alter table `c1de839f-4305-4edc-b6bf-463b5890430c` add column `9b28a0ce-6ee2-4402-89cc-149eabfc2bcb` timestamp DEFAULT '2014-09-12 01:52:09'")
	tk.MustExec("alter table `c1de839f-4305-4edc-b6bf-463b5890430c` alter column `9b28a0ce-6ee2-4402-89cc-149eabfc2bcb` set default '2014-09-13 01:52:09'")
	tk.MustExec("alter table `c1de839f-4305-4edc-b6bf-463b5890430c` add index `4e211d2b-9446-470a-bef8-6842595cdf11` (`31a80f0f-4b8c-4171-9f06-8d191842541f`,`a9b6435c-43a2-4370-9256-eb8b64a590eb`,`de3381af-3319-4035-985e-a83296fbe929`,`ffbf71c5-8559-40f1-a1f8-5e5e03e69abd`,`c37f63a0-7802-48bb-ba0b-7b4c5e2b8097`,`54bac6a1-ada5-436a-8c13-bc6ab738db46`,`9b28a0ce-6ee2-4402-89cc-149eabfc2bcb`)")
	tk.MustExec("alter table `c1de839f-4305-4edc-b6bf-463b5890430c` add index `62b5ddf9-655f-41b0-be91-9e48755728e3` (`9b28a0ce-6ee2-4402-89cc-149eabfc2bcb`)")
	tk.MustExec("set sql_mode=''")
	tk.MustExec("insert into `c1de839f-4305-4edc-b6bf-463b5890430c` values (default,7568.9873,'omp3m',-17188,'06:10:55',41621.6000000000000000000000,6475468268162293516,'HOr#lf71'),('2018-07-13 16:00:00',3412.9521,'ky8',31545,'06:10:55',41621.6000000000000000000000,6475468268162293516,'~0$~ymCy0#ByW@s')")

	tk.MustExec("INSERT INTO `c1de839f-4305-4edc-b6bf-463b5890430c` VALUES ('2011-04-01 16:00:00',7568.9873,'omp3m',-17188,'06:10:55',41621.6000000000000000000000,6475468268162293516,'HOr#lf71'),('2018-07-13 16:00:00',3412.9521,'ky8',31545,'06:10:55',41621.6000000000000000000000,6475468268162293516,'~0$~ymCy0#ByW@s'),('2020-11-07 16:00:00',802.6281,'f',-13254,'19:55:37',61.0000000000000000000000,2887117290787331452,'k('),('2012-10-11 16:00:00',NULL,'ky8',21801,'07:48:31',19199.0000000000000000000000,3585948499425187413,'CotgS15'),('1975-12-24 16:00:00',7932.282,'omp3m',-31167,'06:10:55',41621.6000000000000000000000,6475468268162293516,'vpYJJU56kgeuC9'),('2015-06-22 16:00:00',1.0469099,'s059b',-4305,'18:18:50',83.2130000000000000000000,-6760746055775614715,'uv9qaoowi3d2w5po2d5224fm4rmp5o5turz1db1iimcyv8fu8b5kyturpwvvat73ty8oi9qnp76pres1drv9'),('2019-02-26 16:00:00',202.99762,'b',9030,'06:10:55',41621.6000000000000000000000,6475468268162293516,'HJhoDRb~'),('2006-10-25 16:00:00',1080.9202,'nt',2499,'06:53:28',623985.8425000000000000000000,7865364920877895055,'Bfe5iP&B#iYLU'),('2011-04-01 16:00:00',2699.405,'37n5',31159,'06:10:55',41621.6000000000000000000000,6475468268162293516,'HOr#lf71'),('2011-04-01 16:00:00',4898.466,'ky8',-30725,'06:10:55',41621.6000000000000000000000,6475468268162293516,'HOr#lf71'),('2011-04-01 16:00:00',9216.521,'b',6252,'06:10:55',41621.6000000000000000000000,6475468268162293516,'HOr#lf71'),('2034-09-03 16:00:00',1.0469099,NULL,-6900,'13:23:10',249211.0643000000000000000000,-1445480886636692245,'uv9qaoowi3d2w5po2d5224fm4rmp5o5turz1db1iimcyv8fu8b5kyturpwvvat73ty8oi9qnp76pres1drv9'),('2011-05-20 16:00:00',NULL,'s059b',-30406,'06:10:55',41621.6000000000000000000000,6475468268162293516,'-(^g^1#gxBTd'),('2011-04-01 16:00:00',6854.9395,'1sv7l',-10093,'06:10:55',41621.6000000000000000000000,6475468268162293516,'HOr#lf71'),('2011-04-01 16:00:00',6124.435,'37n5',-10093,'06:10:55',41621.6000000000000000000000,6475468268162293516,'HOr#lf71'),('2011-04-01 16:00:00',1320.1642,'omp3m',-10093,'06:10:55',41621.6000000000000000000000,6475468268162293516,'HOr#lf71'),('2011-04-01 16:00:00',390.43777,'37n5',-10093,'06:10:55',41621.6000000000000000000000,6475468268162293516,'HOr#lf71'),('2011-04-01 16:00:00',4388.9893,NULL,-10093,'06:10:55',41621.6000000000000000000000,6475468268162293516,'HOr#lf71'),('2011-04-01 16:00:00',7885.4834,NULL,-10093,'06:10:55',41621.6000000000000000000000,6475468268162293516,'HOr#lf71'),('2014-09-13 01:52:09',622.371,'1sv7l',-10093,'06:10:55',41621.6000000000000000000000,6475468268162293516,'uv9qaoowi3d2w5po2d5224fm4rmp5o5turz1db1iimcyv8fu8b5kyturpwvvat73ty8oi9qnp76pres1drv9'),('2014-09-13 01:52:09',699.61615,'1sv7l',-10093,'08:08:53',36730.8000000000000000000000,-8643344492354284347,'uv9qaoowi3d2w5po2d5224fm4rmp5o5turz1db1iimcyv8fu8b5kyturpwvvat73ty8oi9qnp76pres1drv9'),('2014-09-13 01:52:09',9382.332,'1sv7l',-10093,'06:10:55',41621.6000000000000000000000,6475468268162293516,'uv9qaoowi3d2w5po2d5224fm4rmp5o5turz1db1iimcyv8fu8b5kyturpwvvat73ty8oi9qnp76pres1drv9'),('2014-09-13 01:52:09',1.0469099,'ky8',-10093,'08:04:07',84447970762009522232772614140425.0705348716771117376820,6957149600293861461,'uv9qaoowi3d2w5po2d5224fm4rmp5o5turz1db1iimcyv8fu8b5kyturpwvvat73ty8oi9qnp76pres1drv9'),('1999-02-16 16:00:00',1.0469099,'1sv7l',1070,'21:22:30',0.7000000000000000000000,6241710034054524186,'uv9qaoowi3d2w5po2d5224fm4rmp5o5turz1db1iimcyv8fu8b5kyturpwvvat73ty8oi9qnp76pres1drv9'),('1988-03-19 16:00:00',1.0469099,NULL,-31877,'09:32:28',5.3000000000000000000000,287263969028687564,'uv9qaoowi3d2w5po2d5224fm4rmp5o5turz1db1iimcyv8fu8b5kyturpwvvat73ty8oi9qnp76pres1drv9'),('1971-10-11 16:00:00',498.11197,'ky8',8350,'06:10:55',41621.6000000000000000000000,6475468268162293516,'HwwcM%rhoUsw6+7Nr@'),('2011-04-01 16:00:00',7930.4033,'1sv7l',11003,'06:10:55',41621.6000000000000000000000,6475468268162293516,'HOr#lf71'),('1993-12-28 16:00:00',6513.789,'s059b',-16666,'06:10:55',41621.6000000000000000000000,6475468268162293516,'%1'),('2011-04-01 16:00:00',3455.8967,'b',14419,'06:10:55',41621.6000000000000000000000,6475468268162293516,'HOr#lf71');")
	tk.MustExec("insert into `c1de839f-4305-4edc-b6bf-463b5890430c` select * from `c1de839f-4305-4edc-b6bf-463b5890430c`")
	tk.MustExec("insert into `c1de839f-4305-4edc-b6bf-463b5890430c` select * from `c1de839f-4305-4edc-b6bf-463b5890430c`")
	tk.MustExec("insert into `c1de839f-4305-4edc-b6bf-463b5890430c` select * from `c1de839f-4305-4edc-b6bf-463b5890430c`")
	tk.MustExec("insert into `c1de839f-4305-4edc-b6bf-463b5890430c` select * from `c1de839f-4305-4edc-b6bf-463b5890430c`")
	tk.MustExec("insert into `c1de839f-4305-4edc-b6bf-463b5890430c` select * from `c1de839f-4305-4edc-b6bf-463b5890430c`")

	tk.MustExec("CREATE TABLE `3fec2a59-2ca9-44cd-a0ef-b7f794c25edb` (\n  `6032b4b1-cffd-469d-b250-5b841b2766c8` date DEFAULT '1603-03-09',\n  `1af1b270-5ce1-49b5-a643-f00851108192` varchar(28) DEFAULT 'gamkppzbtzsgcm31aw2bi',\n  `bf750dda-bfe2-47e4-a047-53b20bedfa93` text DEFAULT NULL\n) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin COMMENT='70b0c58a-8828-4ef1-b26f-22f3d162ab00';")

	tk.MustExec("CREATE TABLE `b29b41f2-48a2-4479-bcac-f381127f91c2` (\n  `5c3d8585-58e2-4753-9b04-2ff275e0f579` tinyblob DEFAULT NULL,\n  `67c8632d-af17-4d83-81de-28a773ce06ec` varchar(245) DEFAULT 'qfoqad38mje03ztxvdvwp2acz3',\n  KEY `f822d170-d530-45ac-9acc-45eaea990ef7` (`67c8632d-af17-4d83-81de-28a773ce06ec`),\n  KEY `fc8f4dba-68aa-4e9b-b0ab-7cc5d7a67e0e` (`67c8632d-af17-4d83-81de-28a773ce06ec`)\n) ENGINE=InnoDB DEFAULT CHARSET=latin1 COLLATE=latin1_bin /*T! SHARD_ROW_ID_BITS=4 */")
	tk.MustExec("alter table `b29b41f2-48a2-4479-bcac-f381127f91c2` add column `1a201f4f-2f69-4aea-bb0a-ef3c9a6adaae` timestamp DEFAULT '2028-01-03 17:04:55'")
	tk.MustExec("alter table `b29b41f2-48a2-4479-bcac-f381127f91c2` add index `3350b038-edb2-4835-b6c5-16b5075e0292` (`1a201f4f-2f69-4aea-bb0a-ef3c9a6adaae`)")
	tk.MustExec("alter table `b29b41f2-48a2-4479-bcac-f381127f91c2` alter `1a201f4f-2f69-4aea-bb0a-ef3c9a6adaae` set DEFAULT '2028-08-03 17:14:55'")

	tk.MustExec("set @@tidb_partition_prune_mode = 'static';")
	tk.MustExec("analyze table `c1de839f-4305-4edc-b6bf-463b5890430c`")
	tk.MustExec("analyze table `3fec2a59-2ca9-44cd-a0ef-b7f794c25edb`")

	for j := 0; j < 10000; j++ {
		tk.MustExec("delete from `b29b41f2-48a2-4479-bcac-f381127f91c2`")
		tk.MustExec("insert into `b29b41f2-48a2-4479-bcac-f381127f91c2`(`5c3d8585-58e2-4753-9b04-2ff275e0f579`, `67c8632d-af17-4d83-81de-28a773ce06ec`) values (NULL, 'qfoqad38mje03ztxvdvwp2acz3'), (0x6975327446316943637A4E36487762745070, 'qfoqad38mje03ztxvdvwp2acz3');")

		tk.MustExec(fmt.Sprintf("update `b29b41f2-48a2-4479-bcac-f381127f91c2` set `b29b41f2-48a2-4479-bcac-f381127f91c2`.`67c8632d-af17-4d83-81de-28a773ce06ec`\n= 'IZW9E00t22222cGov(OSL-%d' where not( `b29b41f2-48a2-4479-bcac-f381127f91c2`.`67c8632d-af17-4d83-81de-28a773ce06ec` in ( select `bf750dda-bfe2-47e4-a047-53b20bedfa93` from `3fec2a59-2ca9-44cd-a0ef-b7f794c25edb` where `b29b41f2-48a2-4479-bcac-f381127f91c2`.`67c8632d-af17-4d83-81de-28a773ce06ec` in ( select `9b28a0ce-6ee2-4402-89cc-149eabfc2bcb` from `c1de839f-4305-4edc-b6bf-463b5890430c` use index() where not( `b29b41f2-48a2-4479-bcac-f381127f91c2`.`67c8632d-af17-4d83-81de-28a773ce06ec` < '' ) ) ) ) or not( `b29b41f2-48a2-4479-bcac-f381127f91c2`.`5c3d8585-58e2-4753-9b04-2ff275e0f579` <= '' );", j))
		logutil.BgLogger().Info("update", zap.Int("j", j))
		tk.MustExec("admin check index `b29b41f2-48a2-4479-bcac-f381127f91c2` `3350b038-edb2-4835-b6c5-16b5075e0292`;")
	}
}
```

To make it reproduce easier:
1.  in getColDefaultValue()
```
@@ -559,9 +560,14 @@ func getColDefaultValue(ctx sessionctx.Context, col *model.ColumnInfo, defaultVa
                        // For col.Version = 0, the timezone information of default value is already lost, so use the system timezone as
 the default value timezone.
                        sc.TimeZone = timeutil.SystemLocation()
                        if col.Version >= model.ColumnInfoVersion1 {
+                               logutil.BgLogger().Warn("set to UTC  in")
                                sc.TimeZone = time.UTC
                        }
-                       defer func() { sc.TimeZone = originalTZ }()
+                       defer func() {
+                               time.Sleep(time.Duration(rand.Intn(500)) * time.Millisecond)
+                               sc.TimeZone = originalTZ
+                               logutil.BgLogger().Warn("set to UTC  out")
+                       }()
                }
        }
```

2. in encodeValueDatum()
```
        case types.KindMysqlTime:
                // for mysql datetime, timestamp and date type
                t := d.GetMysqlTime()
+               logutil.BgLogger().Error("encodeValueDatum in")
+               time.Sleep(time.Duration(rand.Intn(50)) * time.Millisecond)
                if t.Type() == mysql.TypeTimestamp && sc != nil && sc.TimeZone != time.UTC {
                        err = t.ConvertTimeZone(sc.TimeZone, time.UTC)
                        if err != nil {
                                return
                        }
                }
+               logutil.BgLogger().Error("encodeValueDatum out")
                var v uint64
```

```
Note: it's necessary to add a timezone column and then modify its default value, so that it has OriginDefaultValue.
```

### 2. What did you expect to see? (Required)
No error.

### 3. What did you see instead (Required)
```
sql:admin check index `b29b41f2-48a2-4479-bcac-f381127f91c2` `3350b038-edb2-4835-b6c5-16b5075e0292`;, [], error stack [executor:8134]data inconsistency in table: b29b41f2-48a2-4479-bcac-f381127f91c2, index: 3350b038-edb2-4835-b6c5-16b5075e0292, col: 1a201f4f-2f69-4aea-bb0a-ef3c9a6adaae, handle: "2882303761517117441", index-values:"KindMysqlTime 2028-08-03 17:14:55" != record-values:"KindMysqlTime 2028-08-04 01:14:55", compare err:<nil>
```
### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #40710

Problem Summary:
As mentioned in #40710, changing the timezone in the statement context may cause data race in getColDefaultValue().

### What is changed and how it works?
Add an explicit timezone argument instead of changing the timezone in the statement context.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
Run the test mentioned in #40710. since this PR completely avoids this data race, there is no need to write a unit test. And the unit test for it is meaningless.
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
fix a timezone data race which may cause wrong row data
```

### CODE REVIEW

**Review by bb7133 — APPROVED** [review]
LGTM

File: expression/helper.go [hunk 1]
```diff
@@ -84,7 +84,7 @@ func getTimeCurrentTimeStamp(ctx sessionctx.Context, tp byte, fsp int) (t types.
 }
 
 // GetTimeValue gets the time value with type tp.
-func GetTimeValue(ctx sessionctx.Context, v interface{}, tp byte, fsp int) (d types.Datum, err error) {
+func GetTimeValue(ctx sessionctx.Context, v interface{}, tp byte, fsp int, explicitTz *time.Location) (d types.Datum, err error) {
```
[the remark this answers is not in the document]
Comment (wjhuang2016) [reply expression/helper.go:87]: fixed, it's strange that test passed:)

**Review by tiancaiamao — APPROVED** [review]

File: types/time.go [hunk 2]
```diff
@@ -1970,8 +1979,9 @@ func parseDateTimeFromNum(sc *stmtctx.StatementContext, num int64) (Time, error)
 // The valid datetime range is from '1000-01-01 00:00:00.000000' to '9999-12-31 23:59:59.999999'.
 // The valid timestamp range is from '1970-01-01 00:00:01.000000' to '2038-01-19 03:14:07.999999'.
 // The valid date range is from '1000-01-01' to '9999-12-31'
-func ParseTime(sc *stmtctx.StatementContext, str string, tp byte, fsp int) (Time, error) {
-	return parseTime(sc, str, tp, fsp, false)
+// explicitTz is used to handle a data race of timeZone, refer to https://github.com/pingcap/tidb/issues/40710. It only works for timestamp now, be careful to used it!
```
Comment (tangenta) [inline types/time.go:1982]:
```suggestion
// explicitTz is used to handle a data race of timeZone, refer to https://github.com/pingcap/tidb/issues/40710. It only works for timestamp now, be careful to use it!
```

**Pull request merged**